### PR TITLE
fix: Move compilation costs to deploy action

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
- Moving compilation costs from function call at the runtime to a deploy action as part of disk compilation cache effort
- Bumping state to 10M accounts and 1M active to better account for trie lookups.

Issue: https://github.com/nearprotocol/NEPs/issues/120

# Test plan:
- CI

### CODE REVIEW

**Comment by olonho** [comment]
Contract caching is implemented, so we shall extend this PR (or create a new one) with new costs.

File: runtime/runtime-params-estimator/estimate.sh [hunk 1]
```diff
@@ -25,8 +25,8 @@ script_dir="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
 cd "${script_dir}"/../../
 cargo run --release --package neard --bin neard -- --home /tmp/data init --chain-id= --test-seed=alice.near --account-id=test.near --fast
 cd "${script_dir}"/../../genesis-tools/genesis-populate
-cargo run --release --package genesis-populate --bin genesis-populate -- --additional-accounts-num=200000 --home /tmp/data
+cargo run --release --package genesis-populate --bin genesis-populate -- --additional-accounts-num=10000000 --home /tmp/data
```
Comment (olonho) [inline runtime/runtime-params-estimator/estimate.sh:28]: How this increase affects execution time of estimator under simulator?

Comment (evgenykuzyakov) [reply runtime/runtime-params-estimator/estimate.sh:28]: It does increase the time to run and increases the cost by 1.5X at least for initial lookup. But it doesn't block the estimation which is fine.

**Review by ailisp — APPROVED** [review]
As we found nothing wrong with this PR, the issue is a long existing one: #3601

**Comment by olonho** [comment]
See https://github.com/near/nearcore/pull/3646 for explicit contract compilation and storing API.

**Comment by evgenykuzyakov** [comment]
I've regenerate the config using this PR with fix from @ailisp

This only thing that is likely wrong is `function_call_cost_per_byte`. We probably should take a look at this later and override it manually using value from `data_receipt_creation_config::cost_per_byte` for now

```json
{
  "storage_amount_per_byte": "90900000000000000000",
  "transaction_costs": {
    "action_receipt_creation_config": {
      "send_sir": 192685562500,
      "send_not_sir": 192685562500,
      "execution": 192685562500
    },
    "data_receipt_creation_config": {
      "base_cost": {
        "send_sir": 36737650937,
        "send_not_sir": 36737650937,
        "execution": 36737650937
      },
      "cost_per_byte": {
        "send_sir": 13726053,
        "send_not_sir": 13726053,
        "execution": 13726053
      }
    },
    "action_creation_config": {
      "create_account_cost": {
        "send_sir": 28259062500,
        "send_not_sir": 28259062500,
        "execution": 28259062500
      },
      "deploy_contract_cost": {
        "send_sir": 48592000000,
        "send_not_sir": 48592000000,
        "execution": 81763775000
      },
      "deploy_contract_cost_per_byte": {
        "send_sir": 6768990,
        "send_not_sir": 6768990,
        "execution": 442241673
      },
      "function_call_cost": {
        "send_sir": 279961875000,
        "send_not_sir": 279961875000,
        "execution": 279961875000
      },
      "function_call_cost_per_byte": {
        "send_sir": 28490,
        "send_not_sir": 28490,
        "execution": 28490
      },
      "transfer_cost": {
        "send_sir": 23671687500,
        "send_not_sir": 23671687500,
        "execution": 23671687500
      },
      "stake_cost": {
        "send_sir": 59179750000,
        "send_not_sir": 59179750000,
        "execution": 59179750000
      },
      "add_key_cost": {
        "full_access_cost": {
          "send_sir": 26177250000,
          "send_not_sir": 26177250000,
          "execution": 26177250000
        },
        "function_call_cost": {
          "send_sir": 28616375000,
          "send_not_sir": 28616375000,
          "execution": 28616375000
        },
        "function_call_cost_per_byte": {
          "send_sir": 8846293,
          "send_not_sir": 8846293,
          "execution": 8846293
        }
      },
      "delete_key_cost": {
        "send_sir": 11641500000,
        "send_not_sir": 11641500000,
        "execution": 11641500000
      },
      "delete_account_cost": {
        "send_sir": 134024187500,
        "send_not_sir": 134024187500,
        "execution": 134024187500
      }
    },
    "storage_usage_config": {
      "num_bytes_account": 100,
      "num_extra_bytes_record": 40
    },
    "burnt_gas_reward": [
      3,
      10
    ],
    "pessimistic_gas_price_inflation_ratio": [
      103,
      100
    ]
  },
  "wasm_config": {
    "ext_costs": {
      "base": 84931279,
      "contract_compile_base": 0,
      "contract_compile_bytes": 0,
      "read_memory_base": 245588112,
      "read_memory_byte": 1266538,
      "write_memory_base": 312951150,
      "write_memory_byte": 907359,
      "read_register_base": 213150650,
      "read_register_byte": 32284,
      "write_register_base": 336142237,
      "write_register_byte": 1266624,
      "utf8_decoding_base": 423985325,
      "utf8_decoding_byte": 101996609,
      "utf16_decoding_base": 565795550,
      "utf16_decoding_byte": 57021582,
      "sha256_base": 895253475,
      "sha256_byte": 7978920,
      "keccak256_base": 1334262200,
      "keccak256_byte": 7089065,
      "keccak512_base": 1325076737,
      "keccak512_byte": 12150350,
      "log_base": 565795550,
      "log_byte": 4358180,
      "storage_write_base": 10582593250,
      "storage_write_key_byte": 21481737,
      "storage_write_value_byte": 9233784,
      "storage_write_evicted_byte": 5886884,
      "storage_read_base": 23958504750,
      "storage_read_key_byte": 11098765,
      "storage_read_value_byte": 2641228,
      "storage_remove_base": 32343327875,
      "storage_remove_key_byte": 23984112,
      "storage_remove_ret_value_byte": 3682421,
      "storage_has_key_base": 22075726500,
      "storage_has_key_byte": 10800202,
      "storage_iter_create_prefix_base": 0,
      "storage_iter_create_prefix_byte": 0,
      "storage_iter_create_range_base": 0,
      "storage_iter_create_from_byte": 0,
      "storage_iter_create_to_byte": 0,
      "storage_iter_next_base": 0,
      "storage_iter_next_key_byte": 0,
      "storage_iter_next_value_byte": 0,
      "touching_trie_node": 6845287071,
      "promise_and_base": 435863298,
      "promise_and_per_promise": 1827444,
      "promise_return": 125771383,
      "validator_stake_base": 303944908800,
      "validator_total_stake_base": 303944908800
    },
    "grow_mem_cost": 1,
    "regular_op_cost": 1159060,
    "limit_config": {
      "max_gas_burnt": 200000000000000,
      "max_gas_burnt_view": 200000000000000,
      "max_stack_height": 16384,
      "initial_memory_pages": 1024,
      "max_memory_pages": 2048,
      "registers_memory_limit": 1073741824,
      "max_register_size": 104857600,
      "max_number_registers": 100,
      "max_number_logs": 100,
      "max_total_log_length": 16384,
      "max_total_prepaid_gas": 300000000000000,
      "max_actions_per_receipt": 100,
      "max_number_bytes_method_names": 2000,
      "max_length_method_name": 256,
      "max_arguments_length": 4194304,
      "max_length_returned_data": 4194304,
      "max_contract_size": 4194304,
      "max_length_storage_key": 4194304,
      "max_length_storage_value": 4194304,
      "max_promises_per_function_call_action": 1024,
      "max_number_input_data_dependencies": 128
    }
  },
  "account_creation_config": {
    "min_allowed_top_level_account_length": 0,
    "registrar_account_id": "registrar"
  }
}
```

**Comment by evgenykuzyakov** [comment]
@ailisp The function call base cost is still pretty high about 0.56T:
```
"function_call_cost": {
        "send_sir": 279961875000,
        "send_not_sir": 279961875000,
        "execution": 279961875000
      },
```

**Comment by evgenykuzyakov** [comment]
Regenerated config with `io_costs_prod` branch merged:
```json
{
  "storage_amount_per_byte": "90900000000000000000",
  "transaction_costs": {
    "action_receipt_creation_config": {
      "send_sir": 201746000000,
      "send_not_sir": 201746000000,
      "execution": 201746000000
    },
    "data_receipt_creation_config": {
      "base_cost": {
        "send_sir": 40085986937,
        "send_not_sir": 40085986937,
        "execution": 40085986937
      },
      "cost_per_byte": {
        "send_sir": 13730423,
        "send_not_sir": 13730423,
        "execution": 13730423
      }
    },
    "action_creation_config": {
      "create_account_cost": {
        "send_sir": 46143187500,
        "send_not_sir": 46143187500,
        "execution": 46143187500
      },
      "deploy_contract_cost": {
        "send_sir": 65622312500,
        "send_not_sir": 65622312500,
        "execution": 98815775000
      },
      "deploy_contract_cost_per_byte": {
        "send_sir": 6799823,
        "send_not_sir": 6799823,
        "execution": 442328761
      },
      "function_call_cost": {
        "send_sir": 280460812500,
        "send_not_sir": 280460812500,
        "execution": 280460812500
      },
      "function_call_cost_per_byte": {
        "send_sir": 27297,
        "send_not_sir": 27297,
        "execution": 27297
      },
      "transfer_cost": {
        "send_sir": 42300687500,
        "send_not_sir": 42300687500,
        "execution": 42300687500
      },
      "stake_cost": {
        "send_sir": 64921875000,
        "send_not_sir": 64921875000,
        "execution": 64921875000
      },
      "add_key_cost": {
        "full_access_cost": {
          "send_sir": 34072687500,
          "send_not_sir": 34072687500,
          "execution": 34072687500
        },
        "function_call_cost": {
          "send_sir": 35180312500,
          "send_not_sir": 35180312500,
          "execution": 35180312500
        },
        "function_call_cost_per_byte": {
          "send_sir": 8952825,
          "send_not_sir": 8952825,
          "execution": 8952825
        }
      },
      "delete_key_cost": {
        "send_sir": 19335500000,
        "send_not_sir": 19335500000,
        "execution": 19335500000
      },
      "delete_account_cost": {
        "send_sir": 156137812500,
        "send_not_sir": 156137812500,
        "execution": 156137812500
      }
    },
    "storage_usage_config": {
      "num_bytes_account": 100,
      "num_extra_bytes_record": 40
    },
    "burnt_gas_reward": [
      3,
      10
    ],
    "pessimistic_gas_price_inflation_ratio": [
      103,
      100
    ]
  },
  "wasm_config": {
    "ext_costs": {
      "base": 84945869,
      "contract_compile_base": 0,
      "contract_compile_bytes": 0,
      "read_memory_base": 246867187,
      "read_memory_byte": 1266539,
      "write_memory_base": 314210512,
      "write_memory_byte": 907359,
      "read_register_base": 214704800,
      "read_register_byte": 32285,
      "write_register_base": 337430625,
      "write_register_byte": 1266625,
      "utf8_decoding_base": 425211375,
      "utf8_decoding_byte": 102000199,
      "utf16_decoding_base": 567254425,
      "utf16_decoding_byte": 57023927,
      "sha256_base": 906176162,
      "sha256_byte": 8011811,
      "keccak256_base": 1335362200,
      "keccak256_byte": 7089192,
      "keccak512_base": 1327109612,
      "keccak512_byte": 12150528,
      "log_base": 567254425,
      "log_byte": 4358794,
      "storage_write_base": 12464670250,
      "storage_write_key_byte": 24445419,
      "storage_write_value_byte": 12311808,
      "storage_write_evicted_byte": 5930933,
      "storage_read_base": 24285134875,
      "storage_read_key_byte": 11065916,
      "storage_read_value_byte": 2629947,
      "storage_remove_base": 34460499375,
      "storage_remove_key_byte": 24123441,
      "storage_remove_ret_value_byte": 3734313,
      "storage_has_key_base": 22305671250,
      "storage_has_key_byte": 10854894,
      "storage_iter_create_prefix_base": 0,
      "storage_iter_create_prefix_byte": 0,
      "storage_iter_create_range_base": 0,
      "storage_iter_create_from_byte": 0,
      "storage_iter_create_to_byte": 0,
      "storage_iter_next_base": 0,
      "storage_iter_next_key_byte": 0,
      "storage_iter_next_value_byte": 0,
      "touching_trie_node": 6938609964,
      "promise_and_base": 435976301,
      "promise_and_per_promise": 1829841,
      "promise_return": 126021073,
      "validator_stake_base": 303944908800,
      "validator_total_stake_base": 303944908800
    },
    "grow_mem_cost": 1,
    "regular_op_cost": 1159095,
    "limit_config": {
      "max_gas_burnt": 200000000000000,
      "max_gas_burnt_view": 200000000000000,
      "max_stack_height": 16384,
      "initial_memory_pages": 1024,
      "max_memory_pages": 2048,
      "registers_memory_limit": 1073741824,
      "max_register_size": 104857600,
      "max_number_registers": 100,
      "max_number_logs": 100,
      "max_total_log_length": 16384,
      "max_total_prepaid_gas": 300000000000000,
      "max_actions_per_receipt": 100,
      "max_number_bytes_method_names": 2000,
      "max_length_method_name": 256,
      "max_arguments_length": 4194304,
      "max_length_returned_data": 4194304,
      "max_contract_size": 4194304,
      "max_length_storage_key": 4194304,
      "max_length_storage_value": 4194304,
      "max_promises_per_function_call_action": 1024,
      "max_number_input_data_dependencies": 128
    }
  },
  "account_creation_config": {
    "min_allowed_top_level_account_length": 0,
    "registrar_account_id": "registrar"
  }
}
```

Function calls are still large maybe due to a requirement to read a cache from trie and deserialize it.
Or maybe we hashing the code binary somewhere which might cause more. Still this is a little large comparing to `noop` method call.

```json
      "function_call_cost": {
        "send_sir": 280460812500,
        "send_not_sir": 280460812500,
        "execution": 280460812500
      },
      "function_call_cost_per_byte": {
        "send_sir": 27297,
        "send_not_sir": 27297,
        "execution": 27297
      },
```

**Comment by evgenykuzyakov** [comment]
With `SirReceipt` fix for function calls:

```json
{
  "storage_amount_per_byte": "90900000000000000000",
  "transaction_costs": {
    "action_receipt_creation_config": {
      "send_sir": 201683812500,
      "send_not_sir": 201683812500,
      "execution": 201683812500
    },
    "data_receipt_creation_config": {
      "base_cost": {
        "send_sir": 42716319062,
        "send_not_sir": 42716319062,
        "execution": 42716319062
      },
      "cost_per_byte": {
        "send_sir": 13676606,
        "send_not_sir": 13676606,
        "execution": 13676606
      }
    },
    "action_creation_config": {
      "create_account_cost": {
        "send_sir": 43630625000,
        "send_not_sir": 43630625000,
        "execution": 43630625000
      },
      "deploy_contract_cost": {
        "send_sir": 64599375000,
        "send_not_sir": 64599375000,
        "execution": 97899387500
      },
      "deploy_contract_cost_per_byte": {
        "send_sir": 6798332,
        "send_not_sir": 6798332,
        "execution": 442306501
      },
      "function_call_cost": {
        "send_sir": 338806812500,
        "send_not_sir": 338806812500,
        "execution": 338806812500
      },
      "function_call_cost_per_byte": {
        "send_sir": 27665,
        "send_not_sir": 27665,
        "execution": 27665
      },
      "transfer_cost": {
        "send_sir": 39024250000,
        "send_not_sir": 39024250000,
        "execution": 39024250000
      },
      "stake_cost": {
        "send_sir": 65946437500,
        "send_not_sir": 65946437500,
        "execution": 65946437500
      },
      "add_key_cost": {
        "full_access_cost": {
          "send_sir": 37934937500,
          "send_not_sir": 37934937500,
          "execution": 37934937500
        },
        "function_call_cost": {
          "send_sir": 36948437500,
          "send_not_sir": 36948437500,
          "execution": 36948437500
        },
        "function_call_cost_per_byte": {
          "send_sir": 8882256,
          "send_not_sir": 8882256,
          "execution": 8882256
        }
      },
      "delete_key_cost": {
        "send_sir": 18956812500,
        "send_not_sir": 18956812500,
        "execution": 18956812500
      },
      "delete_account_cost": {
        "send_sir": 156484187500,
        "send_not_sir": 156484187500,
        "execution": 156484187500
      }
    },
    "storage_usage_config": {
      "num_bytes_account": 100,
      "num_extra_bytes_record": 40
    },
    "burnt_gas_reward": [
      3,
      10
    ],
    "pessimistic_gas_price_inflation_ratio": [
      103,
      100
    ]
  },
  "wasm_config": {
    "ext_costs": {
      "base": 84943716,
      "contract_compile_base": 0,
      "contract_compile_bytes": 0,
      "read_memory_base": 246795812,
      "read_memory_byte": 1266539,
      "write_memory_base": 314303875,
      "write_memory_byte": 907361,
      "read_register_base": 214361325,
      "read_register_byte": 32285,
      "write_register_base": 337434112,
      "write_register_byte": 1266625,
      "utf8_decoding_base": 424673275,
      "utf8_decoding_byte": 102006449,
      "utf16_decoding_base": 566517037,
      "utf16_decoding_byte": 57020063,
      "sha256_base": 905862262,
      "sha256_byte": 8011757,
      "keccak256_base": 1335301512,
      "keccak256_byte": 7089040,
      "keccak512_base": 1326344337,
      "keccak512_byte": 12150451,
      "log_base": 566517037,
      "log_byte": 4358380,
      "storage_write_base": 11818137625,
      "storage_write_key_byte": 24501586,
      "storage_write_value_byte": 12183491,
      "storage_write_evicted_byte": 5913682,
      "storage_read_base": 25392111125,
      "storage_read_key_byte": 10980087,
      "storage_read_value_byte": 2631267,
      "storage_remove_base": 35062440625,
      "storage_remove_key_byte": 23951839,
      "storage_remove_ret_value_byte": 3670623,
      "storage_has_key_base": 23541503250,
      "storage_has_key_byte": 10717153,
      "storage_iter_create_prefix_base": 0,
      "storage_iter_create_prefix_byte": 0,
      "storage_iter_create_range_base": 0,
      "storage_iter_create_from_byte": 0,
      "storage_iter_create_to_byte": 0,
      "storage_iter_next_base": 0,
      "storage_iter_next_key_byte": 0,
      "storage_iter_next_value_byte": 0,
      "touching_trie_node": 7254888892,
      "promise_and_base": 435901577,
      "promise_and_per_promise": 1821236,
      "promise_return": 125848300,
      "validator_stake_base": 303944908800,
      "validator_total_stake_base": 303944908800
    },
    "grow_mem_cost": 1,
    "regular_op_cost": 1159097,
    "limit_config": {
      "max_gas_burnt": 200000000000000,
      "max_gas_burnt_view": 200000000000000,
      "max_stack_height": 16384,
      "initial_memory_pages": 1024,
      "max_memory_pages": 2048,
      "registers_memory_limit": 1073741824,
      "max_register_size": 104857600,
      "max_number_registers": 100,
      "max_number_logs": 100,
      "max_total_log_length": 16384,
      "max_total_prepaid_gas": 300000000000000,
      "max_actions_per_receipt": 100,
      "max_number_bytes_method_names": 2000,
      "max_length_method_name": 256,
      "max_arguments_length": 4194304,
      "max_length_returned_data": 4194304,
      "max_contract_size": 4194304,
      "max_length_storage_key": 4194304,
      "max_length_storage_value": 4194304,
      "max_promises_per_function_call_action": 1024,
      "max_number_input_data_dependencies": 128
    }
  },
  "account_creation_config": {
    "min_allowed_top_level_account_length": 0,
    "registrar_account_id": "registrar"
  }
}
```

**Review by olonho — APPROVED** [review]

**Pull request merged**